File: backup/Depth-first_search(DFS)/c++/08_괄호변환.cpp
// 이것이 취업을 위한 위한 코딩테스트다 346p
// https://programmers.co.kr/learn/courses/30/lessons/60058
// 2

#include <string>
#include <vector>

using namespace std;

int balancer(string p)
{
    int cnt = 0;
    for (int i = 0; i < p.size(); i++) {
        if (p[i] == '(')
            cnt++;
        else
            cnt--;
        if (cnt == 0)
            return i;
    }
}

bool check(string p)
{
    int cnt = 0;
    for (int i = 0; i < p.size(); i++) {
        if (p[i] == '(')
            cnt++;
        else {
            if (cnt == 0)
                return false;
            cnt--;
        }
    }
    return true;
}

string solution(string p) {
    string answer = "";
    if (p == "")
        return answer;
    int idx = balancer(p);
    string u = "", v = "";
    for (int i = 0; i <= idx; i++)
        u += p[i];
    for (int i = idx + 1; i < p.size(); i++)
        v += p[i];

    if (check(u))
        answer = u + solution(v);
    else {
        answer = "(";
        answer += solution(v);
        answer += ")";
        u = u.substr(1, u.size() - 2);
        for (int i = 0; i < u.size(); i++) {
            if (u[i] == '(')
                u[i] = ')';
            else
                u[i] = '(';
        }
        answer += u;
    }
    return answer;
}

File: backup/Depth-first_search(DFS)/python/08_괄호변환.py
# 이것이 취업을 위한 코딩테스트다 346p
# https://programmers.co.kr/learn/courses/30/lessons/60058
# 2

def balance(p):
    cnt = 0
    for i in range(len(p)):
        if p[i] == '(':
            cnt += 1
        else:
            cnt -= 1
        if cnt == 0:
            return i

def check(p):
    cnt = 0
    for i in p:
        if i == '(':
            cnt += 1
        else:
            if cnt == 0:
                return False
            cnt -= 1
    return True

def solution(p):
    answer = ''
    if p == '':
        return answer
    idx = balance(p)

    u = p[:idx + 1]
    v = p[idx + 1:]

    if check(u):
        answer = u + solution(v)
    else:
        answer = '('
        answer += solution(v)
        answer += ')'
        u = list(u[1:-1])
        for i in range(len(u)):
            if u[i] == '(':
                u[i] = ')'
            else:
                u[i] = '('
        answer += "".join(u)
    return answer

print(solution(input()))